[PDPD FE] : Refactored  PaddlePaddle Quantization

Pey-crypto opened the pull request:
Fix for #20687
1. Supports both per-tensor and per-channel quantization using the axis attribute.
2. Uses FakeQuantize for quantize_linear and a simple subtract-multiply operation for dequantize_linear.
3. The HALF_AWAY_FROM_ZERO round mode is ignored, as it's not explicitly handled.

xczhai commented:
Hi @Pey-crypto 
Thanks for your commits. But we can polish this PR for better maintain.
1. please check https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/paddle/src/internal/pass/transform_fakequantize.cpp is needed or not. Theoretically, such pass should be remove since your commit works.
2. Previously, I prepare some unittests for Quantization. Maybe you can import these tests to such PR. It can ensure the motification accuracy. https://github.com/xczhai/openvino/tree/refactor_paddle_slim
BR.

Pey-crypto commented:
@xczhai ,
Currently working on the errors.

xczhai commented:
> @xczhai , Currently working on the errors.

If you have any questions, feel free reach me. I will reply asap.

xczhai commented:
> > > @xczhai , Currently working on the errors.
> > 
> > 
> > If you have any questions, feel free reach me. I will reply asap.
> 
> https://github.com/xczhai/openvino/blob/a839d4c87e6275851f8e08d518d1e0049d382adb/src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py Is this the correct file?. I thought it would be a c++ test file rather than python.

For 

> > > @xczhai , Currently working on the errors.
> > 
> > 
> > If you have any questions, feel free reach me. I will reply asap.
> 
> https://github.com/xczhai/openvino/blob/a839d4c87e6275851f8e08d518d1e0049d382adb/src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py Is this the correct file?. I thought it would be a c++ test file rather than python.



> > > @xczhai , Currently working on the errors.
> > 
> > 
> > If you have any questions, feel free reach me. I will reply asap.
> 
> https://github.com/xczhai/openvino/blob/a839d4c87e6275851f8e08d518d1e0049d382adb/src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py Is this the correct file?. I thought it would be a c++ test file rather than python.

Hi, 
- The test case and paddle api may be out of date. I have attached the reference link in the tests. If the api is lagacy, you can refer to those paddle repo links. https://github.com/xczhai/openvino/blob/a839d4c87e6275851f8e08d518d1e0049d382adb/src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py#L27
- it is acceptable to cover the OP mapping correctness under some paddle model snippet for testing. C++ test is not needed here. You can refer to https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/paddle/tests/test_models/gen_scripts.

Pey-crypto commented:
@xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help.

xczhai commented:
> @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.

Did you execute "it submodule update --init --recursive" before building codes?

Pey-crypto commented:
> > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> 
> Did you execute `git submodule update --init --recursive` before building codes?

yup multiple times, i even cleared the submodules again and tried it again.

xczhai commented:
> > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > 
> > 
> > Did you execute `git submodule update --init --recursive` before building codes?
> 
> yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully

please share the commands you have tried.

Pey-crypto commented:
> > > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > > 
> > > 
> > > Did you execute `git submodule update --init --recursive` before building codes?
> > 
> > 
> > yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully
> 
> please share the commands you have tried.

git submodule update --init --recursive --progress

Pey-crypto commented:
> > > > > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > > > > 
> > > > > 
> > > > > Did you execute `git submodule update --init --recursive` before building codes?
> > > > 
> > > > 
> > > > yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully
> > > 
> > > 
> > > please share the commands you have tried.
> > 
> > 
> > git submodule update --init --recursive --progress cmake --build . --parallel $(nproc) cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_MLAS=OFF -DENABLE_ONEDNN=OFF ..
> 
> You should turn on mlas and onednn though it takes more time to build.

i did also, but then cmake reports mlas and onednn makefiles aren't available. I will try again and get back with the results.

xczhai commented:
> > > > > > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > > > > > 
> > > > > > 
> > > > > > Did you execute `git submodule update --init --recursive` before building codes?
> > > > > 
> > > > > 
> > > > > yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully
> > > > 
> > > > 
> > > > please share the commands you have tried.
> > > 
> > > 
> > > git submodule update --init --recursive --progress cmake --build . --parallel $(nproc) cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_MLAS=OFF -DENABLE_ONEDNN=OFF ..
> > 
> > 
> > You should turn on mlas and onednn though it takes more time to build.
> 
> i did do that also, but then cmake reports mlas and onednn makefiles aren't available. I will try again and get back with the results.

Could you share the CPU platform? I can try to reproduce it locally.

Pey-crypto commented:
> > > > > > > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > > > > > > 
> > > > > > > 
> > > > > > > Did you execute `git submodule update --init --recursive` before building codes?
> > > > > > 
> > > > > > 
> > > > > > yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully
> > > > > 
> > > > > 
> > > > > please share the commands you have tried.
> > > > 
> > > > 
> > > > git submodule update --init --recursive --progress cmake --build . --parallel $(nproc) cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_MLAS=OFF -DENABLE_ONEDNN=OFF ..
> > > 
> > > 
> > > You should turn on mlas and onednn though it takes more time to build.
> > 
> > 
> > i did do that also, but then cmake reports mlas and onednn makefiles aren't available. I will try again and get back with the results.
> 
> Could you share the CPU platform and OS info? I can try to reproduce it locally.

Operating System: Ubuntu 24.04 LTS
Model name:            12th Gen Intel(R) Core(TM) i5-12500H
Architecture: x86-64

Pey-crypto commented:
@xczhai Finally figured out the error, the permissions on those folders were messed up somehow, attaching strace revealed perms were being changed. I guess it was local issue on my end.

Pey-crypto commented:
> Hi @Pey-crypto Thanks for your commits. But we can polish this PR for better maintain.
> 
>     1. please check https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/paddle/src/internal/pass/transform_fakequantize.cpp is needed or not. Theoretically, such pass should be remove since your commit works.
> 
>     2. Previously, I prepare some unittests for Quantization. Maybe you can import these tests to such PR. It can ensure the motification accuracy. https://github.com/xczhai/openvino/tree/refactor_paddle_slim
> 
> 
> BR.

Removed transform_fakquantize as the node is being made in quantize_linear. Removed references from the paddle frontend.

mlukasze commented:
@xczhai please evaluate current state of the PR, enable jenkins pipelines if good enough

Pey-crypto commented:
@xczhai I used structured bindings, which wasn't compliant with C++ 11 standard, I have replaced it with tie.

xczhai commented:
> @xczhai I used structured bindings, which wasn't compliant with C++ 11 standard, I have replaced it with tie.

Yes, you are right. The current implementation should be aligned with C++11.

Pey-crypto commented:
@xczhai , I have fixed the tests. Could you review it?

xczhai commented:
> @xczhai , I have fixed the tests. Could you review it?

OK. I start it.

xczhai commented on src/frontends/paddle/CMakeLists.txt:
```diff
@@ -2,8 +2,9 @@
-add_subdirectory(src)
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../../core/dev_api)
 
+add_subdirectory(src)
```
why need to modify this CMakeList?

Pey-crypto replied:
I wanted to access a utility func in the validation_util header file in the core_dev folder. I modified the file to tell MakeTool the location of the file.

xczhai replied:
@Pey-crypto 
I see. Do you want to call `ov::util::normailize_axis` in the `openvino/core/validation_util.hpp`? If so, `ov::util::normailize_axis` is a very simple function, which is used to correct axis when it is negative. You can implement it like `(value < 0) ? value + max : value;`.
Need not to modify the Makefile and import more dependency.

xczhai commented on src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py:
```diff
@@ -0,0 +1,104 @@
+import paddle
+from save_model import exportModel
+import sys
+
```
```suggestion
from paddle.base.framework import in_dygraph_mode
if paddle.__version__ >= '2.6.0':
    from paddle.base.layer_helper import LayerHelper
else:
    from paddle.fluid.layer_helper import LayerHelper
```

xczhai commented on src/frontends/paddle/tests/test_models/gen_scripts/generate_dequantize_linear.py:
```diff
@@ -0,0 +1,106 @@
+import paddle
+from save_model import exportModel
+import sys
+
```
```suggestion
from paddle.base.framework import in_dygraph_mode
if paddle.__version__ >= '2.6.0':
    from paddle.base.layer_helper import LayerHelper
else:
    from paddle.fluid.layer_helper import LayerHelper
```

xczhai commented on src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py:
```diff
@@ -0,0 +1,104 @@
+
+    # Test: Different bit lengths
+    quant_axis = -1
+    test_quantize_linear("quantize_linear_4bit", input_data, quant_axis, bit_length=4)
```
For now, we don't support 4-bit. can remove this test.

xczhai commented:
@Pey-crypto 
Two things need to do.

1. need to fix the testcase. for example.
```bash
cd //openvino/src/frontends/paddle/tests
python test_models/gen_scripts/generate_ceil.py ///openvino/bin/intel64/Debug/test_model_zoo/paddle_test_models/

cd //openvino/bin/intel64/Debug
./paddle_tests --gtest_filter=PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_ceil_*
```
2. need to check the executable graph and compare the performance.
```bash
cd //openvino/bin/intel64/Debug
./benchmark -m xxx.pdmodel # can measure the perf
./benchmark -m xxx.pdmodel -exec_graph_path "exec_graph.xml"  # can dump the fused graph, which can be open in netron
```

mlukasze commented:
@Pey-crypto will you have a time to address issues highlighted in recent comment?

Pey-crypto commented:
> @Pey-crypto will you have a time to address issues highlighted in recent comment?

Yes will do within this week, there were festivities😁

mlukasze commented:
that's awesome, looking forward for your commits then!

xczhai commented on src/frontends/onnx/frontend/src/op/dequantize_linear.cpp:
Should not modify onnx implementation. ONNX and PDPD are seperate frontend.
